src/HappyConnOpener.cc

```diff
@@ -0,0 +1,875 @@
+    n_tries(tries),
+    ranOutOfTimeOrAttemptsEarlier_(nullptr)
+{
+    assert(destinations);
+    assert(dynamic_cast<Answer*>(callback_->getDialer()));
+}
+
+HappyConnOpener::~HappyConnOpener()
+{
+    safe_free(host_);
+    delete lastError;
+}
+
+void
+HappyConnOpener::setHost(const char *h)
+{
+    safe_free(host_);
+    if (h)
+        host_ = xstrdup(h);
+}
+
+void
+HappyConnOpener::start()
+{
+    destinations->notificationPending = false;
+    checkForNewConnection();
+}
+
+bool
+HappyConnOpener::doneAll() const
+{
+    if (!callback_)
+        return true; // (probably found a good path and) informed the requestor
+
+    // TODO: Expose AsyncCall::canFire() instead so that code like this can
+    // detect gone initiators without the need to explicitly cancel callbacks.
+    if (callback_->canceled())
+        return true; // the requestor is gone or has lost interest
+
+    if (prime || spare)
+        return false;
+
+    if (ranOutOfTimeOrAttempts())
+        return true; // trying new connection paths prohibited
+
+    if (destinations->empty() && destinations->destinationsFinalized)
+        return true; // there are no more paths to try
+
+    return false;
+}
+
+void
+HappyConnOpener::swanSong()
+{
+    debugs(17, 5, this);
+
+    if (callback_ && !callback_->canceled())
+        sendFailure();
+
+    if (spareWaiting)
+        cancelSpareWait("HappyConnOpener object destructed");
+
+    // TODO: Find an automated, faster way to kill no-longer-needed jobs.
+
+    if (prime) {
+        if (prime.connector)
+            prime.connector->cancel("HappyConnOpener object destructed");
+        prime.clear();
+    }
+
+    if (spare) {
+        if (spare.connector)
+            spare.connector->cancel("HappyConnOpener object destructed");
+        spare.clear();
+        if (gotSpareAllowance) {
+            TheSpareAllowanceGiver.jobDroppedAllowance();
+            gotSpareAllowance = false;
+        }
+    }
+
+    AsyncJob::swanSong();
+}
+
+const char *
+HappyConnOpener::status() const
+{
+    static MemBuf buf;
+    buf.reset();
+
+    buf.append(" [", 2);
+    if (stopReason)
+        buf.appendf("Stopped, reason:%s", stopReason);
+    if (prime) {
+        if (prime.path && prime.path->isOpen())
+            buf.appendf(" prime FD %d", prime.path->fd);
+        else if (prime.connector)
+            buf.appendf(" prime call%ud", prime.connector->id.value);
+    }
+    if (spare) {
+        if (spare.path && spare.path->isOpen())
+            buf.appendf(" spare FD %d", spare.path->fd);
+        else if (spare.connector)
+            buf.appendf(" spare call%ud", spare.connector->id.value);
+    }
+    if (n_tries)
+        buf.appendf(" tries %d", n_tries);
+    buf.appendf(" %s%u]", id.prefix(), id.value);
+    buf.terminate();
+
+    return buf.content();
+}
+
+/// Create "503 Service Unavailable" or "504 Gateway Timeout" error depending
+/// on whether this is a validation request. RFC 2616 says that we MUST reply
```

Comment:
Please refer to RFC 723x requirements now. I do not insist on this change.

Reply:
Fixed in commit 53592b